### PR TITLE
internal/provisioner: add support to NodePortServicePublishingType

### LINKED ISSUE #4499: gateway provisioner: when using a NodePortService, Gateway's port numbers should define nodeport numbers
**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
- Installed the provisioner: 
  ```bash
  kubectl apply -f examples/gateway-provisioner/
  ```
- Created a GatewayClass with the envoy network publishing type set to `NodePortService`:
  ```yaml
  ---
  kind: GatewayClass
  apiVersion: gateway.networking.k8s.io/v1alpha2
  metadata:
    name: contour
  spec:
    controllerName: projectcontour.io/gateway-controller
    parametersRef:
      group: projectcontour.io
      kind: ContourDeployment
      namespace: projectcontour
      name: contour-params
  ---
  kind: ContourDeployment
  apiVersion: projectcontour.io/v1alpha1
  metadata:
    namespace: projectcontour
    name: contour-params
  spec:
    envoy:
      networkPublishing:
        type: NodePortService
  ```
- created a Gateway using that GatewayClass:
  ```yaml
  kind: Gateway
  apiVersion: gateway.networking.k8s.io/v1alpha2
  metadata:
    name: gateway-1
    namespace: projectcontour
  spec:
    gatewayClassName: contour
    listeners:
      - name: http
        protocol: HTTP
        port: 30080
        allowedRoutes:
          namespaces:
            from: All
  ```
- observed that the provisioned Envoy service does *not* have a NodePort value of `30080`, instead it's used for the service port value:
  ```bash
  kubectl -n projectcontour get service/envoy-gateway-1 -o yaml
  ```

**What did you expect to happen:**
When a GatewayClass is parameterized to provision envoy services of type NodePort, I think it's logical that the Gateway's port values should be used for NodePort values.

In this scenario, it probably makes sense to default the service port values to 80/443.

I guess one challenge with this approach is that, because port numbers are required on Gateway listeners, this means you can't get auto-allocated NodePort values, which might be undesirable for some users. There is an open issue upstream (https://github.com/kubernetes-sigs/gateway-api/issues/1061) to allow port numbers to be optional, which would be the best of both worlds here.

Perhaps we should hold changing this until that issue has landed.

### LINKED ISSUE #4724: Tracking usability enhancements to ContourDeployment CRD
To make the Contour Gateway Provisioner and ContourDeployment more useful, more fields are required to customize the resources it generates, particularly the Contour Deployment, Envoy DaemonSet/Deployment, and Envoy Service.

Let's use this issue to collect the work in progress and organize how we want to develop the API so we can make changes in a consistent way.

The existing API looks like (godoc comments removed for brevity and relevant sub-types included):

```
type ContourDeploymentSpec struct {
	// +optional
	Contour *ContourSettings `json:"contour,omitempty"`

	// +optional
	Envoy *EnvoySettings `json:"envoy,omitempty"`

	// +optional
	RuntimeSettings *ContourConfigurationSpec `json:"runtimeSettings,omitempty"`
}

type ContourSettings struct {
	// +kubebuilder:validation:Minimum=0
	Replicas int32 `json:"replicas,omitempty"`

	// +optional
	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`

	// +kubebuilder:validation:Minimum=0
	// +kubebuilder:validation:Maximum=9
	// +optional
	KubernetesLogLevel uint8 `json:"kubernetesLogLevel,omitempty"`

	// +optional
	LogLevel LogLevel `json:"logLevel,omitempty"`
}

type EnvoySettings struct {
	// +optional
	WorkloadType WorkloadType `json:"workloadType,omitempty"`

	// +kubebuilder:validation:Minimum=0
	Replicas int32 `json:"replicas,omitempty"`

	// +optional.
	NetworkPublishing *NetworkPublishing `json:"networkPublishing,omitempty"`

	// +optional
	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
}

type NetworkPublishing struct {
	// +optional
	Type NetworkPublishingType `json:"type,omitempty"`

	// +optional
	ServiceAnnotations map[string]string `json:"serviceAnnotations,omitempty"`
}

type NodePlacement struct {
	// +optional
	NodeSelector map[string]string `json:"nodeSelector,omitempty"`

	// +optional
	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
}
```

From the open PRs it seems we may be eventually replicating most if not all configurable fields on Services, Deployments/DaemonSets

Relevant PRs:
- [x] Logging:
  - https://github.com/projectcontour/contour/pull/4676
  - https://github.com/projectcontour/contour/pull/4677
  - https://github.com/projectcontour/contour/pull/4801
- [x] Resource labels
  - https://github.com/projectcontour/contour/pull/4709
- [x] Resource annotations
  - https://github.com/projectcontour/contour/pull/4681
- [x] Resource requirements for Envoy/Contour pods
  - https://github.com/projectcontour/contour/pull/4712
- [x] Update Strategy
  - https://github.com/projectcontour/contour/pull/4713
- [x] Extra volumes/mounts
  - https://github.com/projectcontour/contour/pull/4680
- [x] Extra Envoy Service configuration
  - https://github.com/projectcontour/contour/pull/4973
  - https://github.com/projectcontour/contour/pull/4803
  - LoadBalancerIP?
  - Other Service configuration?

### PR DESCRIPTION
**Pull request by izturn**
Signed-off-by: Gang Liu <gang.liu@daocloud.io>

Add support to `NodePortServicePublishingType`

fix #4499 
fix #4724

### CODE REVIEW

File: changelogs/unreleased/4973-izturn-small.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Add support to `NodePortServicePublishingType`. fix [#4499](https://github.com/projectcontour/contour/issues/4499).
```
Comment (skriss) [inline changelogs/unreleased/4973-izturn-small.md:1]:
```suggestion
Gateway API: when provisioning an Envoy `NodePortService`, use the Listeners' port numbers to populate the Service's node port values.
```

**Review by skriss — CHANGES_REQUESTED** [review]

**Review by skriss — APPROVED** [review]
LGTM, thanks @izturn!

**Comment by izturn** [comment]
> Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)
> 
> Add support to `NodePortServicePublishingType`
> 
> fix #4499 fix #4724

@skriss maybe we need to do some adjust to `ContourDeployment` for `LoadBalancer`, so it would better to keep the #4724 open.

**Comment by skriss** [comment]
> > Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)
> > Add support to `NodePortServicePublishingType`
> > fix #4499 fix #4724
> 
> @skriss maybe we need to do some adjust to `ContourDeployment` for `LoadBalancer`, so it would better to keep the #4724 open.

That issue was helpful for organizing the initial wave of PRs but now that everything on there has merged, I think we can close it out. Please open new issues for additional features!

File: apis/projectcontour/v1alpha1/contourdeployment.go [hunk 2]
```diff
@@ -234,6 +234,12 @@ type NetworkPublishing struct {
 	//
 	// See: https://kubernetes.io/docs/concepts/services-networking/service/#nodeport
 	//
+	// NOTE:
+	// When provisioning an Envoy `NodePortService`, use the Listeners' port numbers to populate
```
Comment (sunjayBhatia) [inline apis/projectcontour/v1alpha1/contourdeployment.go:238]: nit: to make it clear we're talking about the Gateway Listener configuration not an Envoy Listener
```suggestion
	// When provisioning an Envoy `NodePortService`, use Gateway Listener port numbers to populate
```

**Review by sunjayBhatia — COMMENTED** [review]
tried this out manually yesterday and it looks good, just one nit comment on some wording

**Comment by izturn** [comment]
> > > Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)
> > > Add support to `NodePortServicePublishingType`
> > > fix #4499 fix #4724
> > 
> > 
> > @skriss maybe we need to do some adjust to `ContourDeployment` for `LoadBalancer`, so it would better to keep the #4724 open.
> 
> That issue was helpful for organizing the initial wave of PRs but now that everything on there has merged, I think we can close it out. Please open new issues for additional features!

got it

**Review by skriss — APPROVED** [review]
LGTM, leaving for @sunjayBhatia

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**